File: ios/Sources/EdgeFrameworks/Core/EdgeVectorStore.swift
public protocol EdgeVectorStore: Sendable {
    func upsert(
        chunks: [EdgeChunk],
        embeddings: [EdgeEmbedding]
    ) async throws

    func search(
        query: EdgeEmbedding,
        topK: Int,
        filter: EdgeVectorFilter?
    ) async throws -> [EdgeSearchResult]

    func remove(
        filter: EdgeVectorFilter
    ) async throws -> Int

    func removeAll() async
}

public extension EdgeVectorStore {
    func search(
        query: EdgeEmbedding,
        topK: Int
    ) async throws -> [EdgeSearchResult] {
        try await search(
            query: query,
            topK: topK,
            filter: nil
        )
    }
}

File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeVectorStore.kt
package io.github.rajibsarwar.edgeframeworks

interface EdgeVectorStore {
    suspend fun upsert(
        chunks: List<EdgeChunk>,
        embeddings: List<EdgeEmbedding>
    )

    suspend fun search(
        query: EdgeEmbedding,
        topK: Int,
        filter: EdgeVectorFilter? = null
    ): List<EdgeSearchResult>

    suspend fun remove(
        filter: EdgeVectorFilter
    ): Int

    suspend fun removeAll()
}
